[HUDI-2161] Adding support to disable meta columns with bulk insert operation

nsivabalan opened the pull request:
## What is the purpose of the pull request

- Adding support to disable meta columns with bulk insert operation in row writer path. For use-cases involving mostly append_only requirements (event logs), users don't have much necessity to do any updates as such. And if such users don't really require incremental queries, we can disable all meta columns with hudi table. And this patch adds such support to bulk_insert operation. 
- Other operations such as update, delete, compaction, clustering etc will be added in future patches. This patch adds support only for bulk_insert. But still queryable though. 
- All built in key generators are supported in this flow (bulk insert). 
- Few constraints in this flow (disabling meta columns with bulk_insert):
  - No Sort Modes are supported. 
  - No incremental support. 

New config introduced: "hoodie.populate.meta.columns" whose default value is true. 
This config value will be serialized to hoodie.properties. Once disabled, it can never be enabled back for a given hudi table. But if users have an existing hudi table, this config can be disabled after they upgrade w/o needing any extra upgrade steps. 
- Added a new api to BuiltInKeyGenerator to fetch partition path from InternalRow. 

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

I am yet to add some more tests for some of the changes in this patch. But have verified locally via spark shell for non partitioned, partitioned w/ simple and complex key gen. Wanted to open up for reviews as I add more tests.

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

nsivabalan commented on hudi-spark-datasource/hudi-spark/src/main/java/org/apache/hudi/HoodieDatasetBulkInsertHelper.java:
```diff
@@ -110,6 +151,55 @@
+   * @param rows
+   * @return
+   */
+  public static Dataset<Row> prepareHoodieDatasetForBulkInsertAppendOnly(Dataset<Row> rows, HoodieWriteConfig config, boolean isGlobalIndex) {
```
this method is not used. both flows use the previous method.

nsivabalan commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala:
```diff
@@ -377,6 +397,63 @@ object HoodieSparkSqlWriter {
     (syncHiveSuccess, common.util.Option.ofNullable(instantTime))
   }
 
+  def bulkInsertAppendOnlyAsRow(sqlContext: SQLContext,
```
bulkInsertAsRow and bulkInsertAppendOnlyAsRow has lot of common code. I am yet to unify them.

nsivabalan commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala:
```diff
@@ -377,6 +398,60 @@ object HoodieSparkSqlWriter {
     (syncHiveSuccess, common.util.Option.ofNullable(instantTime))
   }
 
+  def bulkInsertAsRowNoMetaColumns(sqlContext: SQLContext,
```
this has lot of commonality between existing bulk_insert for row writer path. Once I get first set of reviews, will unify them.

nsivabalan replied:
I meant for changes in HoodieSparkSqlwriter. but both uses the same custom datasource.

vinothchandar replied:
lets please unify this.

nsivabalan reviewed (commented):
added few notes wherever applicable.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/keygen/BuiltinKeyGenerator.java:
```diff
@@ -81,7 +99,55 @@ public String getPartitionPath(Row row) {
+            SimpleAnalyzer$.MODULE$);
+  }
+
+  private static Row deserializeRow(ExpressionEncoder encoder, InternalRow row)
```
yet to test this method. I found a similar method for serializeFromRow and came up with this. But have fixed all other build in key gens like simple, complex, timestamp, custom. Will update the patch once I have the fix. Wanted to open up for reviews as I work on them.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/keygen/BuiltinKeyGenerator.java:
```diff
@@ -81,7 +99,18 @@ public String getPartitionPath(Row row) {
+   * @return the partition path.
+   */
+  public String getPartitionPath(InternalRow internalRow, StructType structType) {
+    throw new UnsupportedOperationException("Operation not supported. Please override if required.");
```
Yet to figure out a way to fix this. Will update the patch once I have the solution. But atleast for all built in key gen, have concrete impls.

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala:
```diff
@@ -128,14 +128,35 @@ object HoodieSparkSqlWriter {
         tableConfig = tableMetaClient.getTableConfig
+      } else {
+        // validate table properties
+        val tableMetaClient = HoodieTableMetaClient.builder().setBasePath(path.get).setConf(sparkContext.hadoopConfiguration).build()
```
is there other places where we do such validation. I think this should be done at the WriteClient level or so? not this high up the stack?

vinothchandar replied:
In any case, please pull all this validation into its own method.

nsivabalan replied:
I am thinking to add the validation within HoodieTableMetaClient. Because, a) We don't instantiate WriteClient at all in row writer path as of now. b) table properties are available when tableMetaclient is instantiated. Even if we were to place it within WriteClient, we have to instantiate Metaclient to read the table props. So, we might as well place the validation method in HoodieTableMetaClient.

nsivabalan replied:
Added a private method here in HoodiesparkSqlWriter wrt params and add a method in HoodieTableMetaclient to validate table properties.

nsivabalan replied:
I understand that we need one place to do the validation. 
As of now, any callers directly to WriteClient has to make another call to do validation. 

I could think of another option. Basically fix all startCommit() methods in WriteClient to take in operationType and do the validation. But there are quite a few callers which I need to fix. If we go ahead with this approach, validation happens within our custom data source where we instantiate the writeClient and start the commit (in row writer path). 
Once we have consensus I can make the changes. I feel this is the right approach.

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala:
```diff
@@ -128,14 +128,35 @@ object HoodieSparkSqlWriter {
+
+      if ( !parameters.getOrElse(HoodieTableConfig.HOODIE_POPULATE_META_COLUMNS.key(), HoodieTableConfig.HOODIE_POPULATE_META_COLUMNS.defaultValue()).toBoolean
+        && operation != WriteOperationType.BULK_INSERT) {
+        throw new HoodieException(HoodieTableConfig.HOODIE_POPULATE_META_COLUMNS.key() + " can only be enabled for " + WriteOperationType.BULK_INSERT
```
you mean - can only be disabled for?

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala:
```diff
@@ -128,14 +128,35 @@ object HoodieSparkSqlWriter {
+        val tableMetaClient = HoodieTableMetaClient.builder().setBasePath(path.get).setConf(sparkContext.hadoopConfiguration).build()
+        if (!tableMetaClient.getTableConfig.populateMetaColumns() &&
+          parameters.getOrElse(HoodieTableConfig.HOODIE_POPULATE_META_COLUMNS.key(), HoodieTableConfig.HOODIE_POPULATE_META_COLUMNS.defaultValue()).toBoolean) {
+          throw new HoodieException(HoodieTableConfig.HOODIE_POPULATE_META_COLUMNS.key() + " already disabled for the table. Can't be enabled back. ")
```
cant be re-enabled.

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala:
```diff
@@ -128,14 +128,35 @@ object HoodieSparkSqlWriter {
 
       // short-circuit if bulk_insert via row is enabled.
       // scalastyle:off
+      if (hoodieConfig.getBoolean(ENABLE_ROW_WRITER_OPT_KEY) &&
```
please move the entire row writing block into its own method.

vinothchandar reviewed (commented):
Lot of cosmetic comments and code cleanups. Reviewed the high level logic, seems ok to me.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/keygen/BuiltinKeyGenerator.java:
```diff
@@ -81,7 +105,52 @@ public String getPartitionPath(Row row) {
+  private static ExpressionEncoder getRowEncoder(StructType schema) {
+    List<Attribute> attributes = JavaConversions.asJavaCollection(schema.toAttributes()).stream()
+        .map(Attribute::toAttribute).collect(Collectors.toList());
+    return RowEncoder.apply(schema)
```
yes.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/keygen/SimpleKeyGenerator.java:
```diff
@@ -72,4 +75,16 @@ public String getPartitionPath(Row row) {
+
+  @Override
+  public String getPartitionPath(InternalRow row, StructType structType) {
+    buildFieldDataTypesMapIfNeeded(structType);
```
yes, InternalRow does not have the come with a schema unfortunately.

nsivabalan replied:
During instantiation of KeyGen, we never know if we are ever going to invoke these methods. As you might be aware, these are invoked only in row writer path and only if meta cols are disabled. So, as of now, our constructors for keyGen are are designed in such a way. So, may be we can add public methods to set the structType and expect callers to set it appropriately before calling these methods. 
I took precendence from what we did when we added support for getPartitionPath(Row). 
Let me know how you wanna go about it.

nsivabalan commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java:
```diff
@@ -675,6 +676,11 @@ public PropertyBuilder setBootstrapBasePath(String bootstrapBasePath) {
       return this;
     }
 
+    public PropertyBuilder setPopulateMetaColumns(boolean populateMetaColumns) {
```
sure. will fix everywhere.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/HoodieAppendOnlyRowCreateHandle.java:
```diff
@@ -0,0 +1,65 @@
+ * limitations under the License.
+ */
+
+package org.apache.hudi.io;
```
sure.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/HoodieAppendOnlyRowCreateHandle.java:
```diff
@@ -0,0 +1,65 @@
+  @Override
+  public void write(InternalRow record) throws IOException {
+    try {
+      fileWriter.writeRow("", record);
```
Here is the reason why I did not fix it. As of now we have HoodieInternalRowFileWriter which has writeRow(recordKey, InternalRow). I did not want to introduce another interface bcoz, I can't extend this new RowCreateHandle from existing RowCreateHandle as both will impl diff interfaces. 
But guess, I can add another overloaded write method to the same interface and each class can call into one of them. 
Will fix it

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/storage/HoodieAppendOnlyRowParquetWriteSupport.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Hoodie Write Support for directly writing Row to Parquet.
+ */
+public class HoodieAppendOnlyRowParquetWriteSupport extends HoodieRowParquetWriteSupport {
```
actually existing ParquetWriteSupport will handle null bloom filter. So, removing this class.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/keygen/BuiltinKeyGenerator.java:
```diff
@@ -81,7 +105,52 @@ public String getPartitionPath(Row row) {
+  }
+
+  private ExpressionEncoder getEncoder(StructType structType) {
+    if (encoder == null) {
```
deserializeRow() is a static method and hence.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/keygen/BuiltinKeyGenerator.java:
```diff
@@ -81,7 +105,52 @@ public String getPartitionPath(Row row) {
+  private static Row deserializeRow(ExpressionEncoder encoder, InternalRow row)
+      throws InvocationTargetException, IllegalAccessException, NoSuchMethodException, ClassNotFoundException {
+    // TODO remove reflection if Spark 2.x support is dropped
+    if (package$.MODULE$.SPARK_VERSION().startsWith("2.")) {
```
I could not find any. 
```
grep -irl "import org.apache.spark.sql.catalyst.expressions.Attribute" hudi-*/
hudi-client//hudi-spark-client/src/test/java/org/apache/hudi/testutils/SparkDatasetTestUtils.java
hudi-client//hudi-spark-client/src/test/java/org/apache/hudi/testutils/KeyGeneratorTestUtilities.java
hudi-client//hudi-spark-client/src/main/java/org/apache/hudi/keygen/BuiltinKeyGenerator.java
hudi-integ-test//src/main/java/org/apache/hudi/integ/testsuite/dag/nodes/ValidateDatasetNode.java
hudi-spark-datasource//hudi-spark/src/test/java/org/apache/hudi/TestHoodieDatasetBulkInsertHelper.java
hudi-spark-datasource//hudi-spark/src/main/java/org/apache/hudi/SparkRowWriteHelper.java
hudi-spark-datasource//hudi-spark/src/main/scala/org/apache/spark/sql/hudi/analysis/HoodieAnalysis.scala
```
In most of these places, we do have a static method to getEncode, but the deserializeRow is first of its kind. We do have serializeRow which converts Row to InternalRow in KeyGeneratorTestUtilities, but its converts from Row -> InternalRow and its part of test code.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/HoodieSparkUtils.scala:
```diff
@@ -21,21 +21,24 @@ package org.apache.hudi
 import org.apache.avro.Schema
 import org.apache.avro.generic.GenericRecord
 import org.apache.hadoop.fs.{FileSystem, Path}
+import org.apache.hudi.client.utils.SparkRowSerDe
```
Moved HoodiesparkUtils, SparkAdaptorSupport and SparkAdaptor from hudi-spark module to hudi-spark-client module since we wanted to access SparkAdaptor from within BuiltInKeygen.

nsivabalan reviewed (commented):
Have addressed all feedback.

nsivabalan commented:
Sorry I had squashed all commits to one so that its easier for me to rebase. I had 10+ files in conflict when I rebased w/ master. Also, I did rename quite a few files and moved some of them(reuse spark serDe class), and so went ahead and squashed it. sorry if you were planning to review just the latest commit.
